fedora/client/openidcclient.py

```diff
@@ -0,0 +1,385 @@
+        self.debug('Refreshing cache')
+        if not os.path.isdir(self.cachedir):
+            self.debug('Creating directory')
+            os.makedirs(self.cachedir)
+        if not os.path.exists(self.cachefile):
+            self.debug('Creating file')
+            with open(self.cachefile, 'w') as f:
+                f.write(json.dumps({}))
+        with open(self.cachefile, 'r') as f:
+            self._cache = json.loads(f.read())
+        self.debug('Loaded %i tokens' % len(self._cache))
+
+    def _refresh_cache(self):
+        with self._cache_lock:
+            self.__refresh_cache()
+
+    def __write_cache(self):
+        assert self._cache_lock.locked()
+        self.debug('Writing cache with %i tokens' % len(self._cache))
+        with open(self.cachefile, 'w') as f:
+            f.write(json.dumps(self._cache))
+
+    def _add_token(self, token):
+        uuid = uuidgen().get_hex()
+        self.debug('Adding token %s to cache' % uuid)
+        with self._cache_lock:
+            self.__refresh_cache()
+            self._cache[uuid] = token
+            self.__write_cache()
+        return uuid
+
+    def _delete_token(self, uuid):
+        self.debug('Removing token %s from cache' % uuid)
+        with self._cache_lock:
+            self._refresh_cache()
+            if uuid in self._cache:
+                del self._cache[uuid]
+                self._write_cache()
+
+    def _get_token_with_scopes(self, scopes):
+        possible_token = None
+        self.debug('Trying to get token with scopes %s' % scopes)
+        for uuid in self._cache:
+            self.debug('Checking %s' % uuid)
+            token = self._cache[uuid]
+            if token['idp'] != self.idp:
+                self.debug('Incorrect idp')
+                continue
+            for scope in scopes:
+                if not scope in token['scopes']:
+                    self.debug('Missing scope %s' % scope)
+                    continue
+            if token['expires_at'] < time.time():
+                # This is a token that's supposed to still be valid, prefer it
+                # over any others we have
+                self.debug('Not yet expired, returning')
+                return uuid, token
+            # This is a token that may or may not still be valid
+            self.debug('Possible')
+            possible_token = (uuid, token)
+        if possible_token:
+            self.debug('Returning possible token')
+            return possible_token
+
+    def _idp_url(self, endpoint):
+        return self.idp + endpoint
+
+    def _get_token_info(self, token):
+        self.debug('Retrieving token info')
+        resp = requests.post(self._idp_url('TokenInfo'),
+                             data={'client_id': self.client_id,
+                                   'client_secret': self.client_secret,
+                                   'token': token['access_token']})
+        # TODO: Be a bit more stable in handling if we get a non-JSON object
```

Comment:
I know this is a work-in-progress that you haven't tested and kindly opened for me, but here's some drive-by feedback ✈️ 

I'd recommend doing something like
```
# Throw an exception for a HTTP 400+ code
resp.raise_for_status()
try:
    return resp.json()
except JSONDecodeError:
   raise SomeExceptionThatTellsUsersTheServerShouldHaveGivenUsJSON
```